### ISSUE TITLE
Real setup method for apps

### ISSUE THREAD

**Issue by xsawyerx** [PerlDancer/Dancer2#1320]
One common annoyance I find more and more is that applications cannot declare a setup method to be run prior to creating the app.

You have two options:
1. Run code as soon as the class is loaded: This is problematic because it won't be tied to a caller cleanly and will not receive parameters.
2. Manually have a method called (I wrote about it [here](http://advent.perldancer.org/2014/18)) and hope no one forgot to call it before calling `to_app`.

I suggest creating a proper one, called `prepare_app`, based on the equivalent one in `Plack`.

My suggestion is to first call warning for any such found methods in apps, to make sure that the subroutine is not found or the users know about it. Then after a number of releases or enough time, to call such a method by default, having given any users that have `prepare_app` in their code a fair warning we intend to do it.

What do you think, @PerlDancer?

**Comment by veryrusty** [comment]
#1331 has been in place for 2 years now. Think it is time for `prepare_app` to be a real thing.

I would like to see this in the next release. If anyone wants to have a go, the floor is yours :dancers:

**Comment by xsawyerx** [comment]
Do we have a pending release?

**Comment by cromedome** [comment]
I don't presently have one scheduled, but it would be nice to put one out before TPC::NA.

I just bought a house and am moving this weekend. After I get settled a bit, I will get the ball rolling on one.

Speaking of TPC::NA, any interest in putting together a Dancer BOF, or at the very least, a Dancer Happy Hour? :)

**Comment by xsawyerx** [comment]
I started implementing this and realized there is a problem in the original design of a `prepare_app` subroutine. When you have an App that is based on another app, we need to know which subroutine to call (the parent app or the subclassing app). I think we should run both.

Example:

```perl
package Foo {
  use Dancer2;
  sub prepare_app {...}
}

package Bar {
  use Dancer2 'appname' => 'Foo';
  sub prepare_app {...}
}
```

In either case, the inheritance could be an arbitrary amount of levels, which would require us to somehow either inspect the calling  (which I'm not sure we have here) or we have to start tracking inheritance when people set `appname`.

An alternative is to register a callback:

```perl
package Foo {
  use Dancer2;
  prepare_app {...}; # notice the semicolon at the end
}

package Bar {
  use Dancer2 'appname' => 'Foo';
  prepare_app {...}; # notice the semicolon at the end
}
```

That way we just stack them up and call all of them in the order of registration.

Thoughts?

**Comment by cromedome** [comment]
I agree - we should definitely run both. While I don't have a current use for that, I can sure imagine some situations where that would be useful. The callback sounds like the most attractive way of implementing this.

**Comment by cromedome** [comment]
This is resolved by #1504. Closing!
